pandas/core/reshape/merge.py

```diff
@@ -2407,13 +2408,16 @@ def _factorize_keys(
                 or is_string_dtype(lk.dtype)
                 and not sort
             )
+            or (isinstance(lk.dtype, StringDtype) and lk.dtype.storage == "pyarrow")
```

Comment:
the compound if statement here is non-obvious.  can it be de-compounded by moving the pyarrow/string case up above this one?

Reply:
Yeah not a problem.

Good to merge then?

Reply:
do we have tests that hit the affected cases?  if so then yes.

Reply:
Yep, ``test_merge_string_dtype`` for example